### PR TITLE
[7.0.x] intermediate upgrades

### PR DESCRIPTION
**Pull request by a-palchikov**
## Description

This PR ports support for intermediate upgrade paths from 5.5.x branch. I've removed the ability to generate the planet/teleport configuration packages as well as planet secrets using CLI interface of the intermediate gravity binary as it appears to be unnecessary for 7.0.x (when going from 5.5.x via 6.1.x). See a comment below for additional thoughts/concerns.

## Type of change

* New feature (non-breaking change which adds functionality)
* This change has a user-facing impact

## Linked tickets and other PRs

* Refs https://github.com/gravitational/gravity/issues/1588

* Requires https://github.com/gravitational/gravity.e/pull/4320, https://github.com/gravitational/planet/pull/714

* Ports https://github.com/gravitational/gravity/pull/637

## TODOs

- [x] Self-review the change
- [x] Write tests
- [x] Perform manual testing
- [ ] Write documentation
- [ ] Address review feedback

### 1. 3-node cluster (1 controller, 2 workers)
    1. Install `5.5.49` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation includes an intermediate hop via `6.1.31`
    1. Upgrade to this branch

### 2. 3-node cluster (3 controllers)
    1. Install `5.5.49` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation includes an intermediate hop via `6.1.31`
    1. Upgrade to this branch

### 3. 3-node cluster (1 controller, 2 workers)
    1. Install `6.1.30` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation does not include any intermediate hops and goes directly to this branch

### 3-node cluster (3 controllers)
    1. Install `6.1.30` 
    1. Upload upgrade for this branch and start the operation in manual mode
    1. Make sure that the upgrade operation does not include any intermediate hops and goes directly to this branch

## Additional information

While working on the port, it occurred to me that in order to be truly generic w.r.t intermediate hops we need to be able to generate sub-plans for each intermediate hops and have the binary of any given version execute it both ways during the plan execution. I did not explore the idea too far but I feel this would be the next logical iteration that will enable co-operation of arbitrary versions without the fear that something might break when only the newest version is doing all the steps (as the case is now).

### CODE REVIEW

File: lib/app/sync.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	// FieldLogger specifies the logger
+	log.FieldLogger
+	// PackService specifies the package service
+	PackService pack.PackageService
+	// AppService specifies the application service
+	AppService Applications
+	// ImageService specifies the docker registry service
+	ImageService docker.ImageService
+	// Progress specifies the optional progress message printer
+	Progress utils.Printer
```
Comment (a-palchikov) [inline lib/app/sync.go:97]: I removed the `ScanConfig` attribute since it was not being used anywhere.

File: lib/fsm/kubernetes.go [hunk 2]
```diff
@@ -104,6 +105,21 @@ func GetUpsertBootstrapResourceFunc(client *kubernetes.Clientset) resources.Reso
 				return trace.Wrap(rigging.ConvertError(err))
 			}
 			log.Debugf("Updated PodSecurityPolicy %q.", resource.Name)
+		case *v1beta1.PodSecurityPolicy:
```
Comment (a-palchikov) [inline lib/fsm/kubernetes.go:108]: This has been reverted to support RBAC configurations from older versions when using intermediate hops.

File: lib/loc/loc.go [hunk 3]
```diff
@@ -238,6 +234,24 @@ func Deduplicate(ls []Locator) (result []Locator) {
 	return result
 }
 
+// GetUpdatedDependencies compares installedDeps against the updateDeps
+// and returns only locators from updateDeps that are updates of those given with installedDeps
+func GetUpdatedDependencies(installedDeps, updateDeps []Locator) ([]Locator, error) {
```
Comment (a-palchikov) [inline lib/loc/loc.go:239]: Moved from lib/app from what used to be [GetUpdatedDependencies](https://github.com/gravitational/gravity/blob/1d5bd035e9cddefafaf5d9f1471f68e9589a3d45/lib/app/apps.go#L141)

File: lib/ops/constants.go [hunk 4]
```diff
@@ -241,4 +243,11 @@ var (
 		OperationUpdateConfig:         SiteStateUpdatingConfig,
 		OperationReconfigure:          SiteStateFailed,
 	}
+
+	// TeleportBrokenJoinTokenVersion is version of the release affected by
+	// the issue with Teleport using incorrect auth token on joined nodes.
+	//
+	// Github issue: https://github.com/gravitational/gravity/issues/1445.
+	// KB: https://community.gravitational.com/t/recover-teleport-nodes-failing-to-join-due-to-bad-token/649.
+	TeleportBrokenJoinTokenVersion = semver.New("5.5.40")
```
Comment (a-palchikov) [inline lib/ops/constants.go:252]: Moved from [here](https://github.com/gravitational/gravity/blob/1d5bd035e9cddefafaf5d9f1471f68e9589a3d45/lib/ops/opsservice/versions.go#L264) which now has been moved [here](https://github.com/gravitational/gravity/tree/dmitri/7.0.x/1588-intermediate-upgrades/lib/update/cluster/versions)

File: lib/ops/ops.go [hunk 5]
```diff
@@ -1106,16 +1106,7 @@ type SiteOperations []storage.SiteOperation
...
-		return s.InstallExpand.Vars
-	}
-	if s.Shrink != nil {
-		return s.Shrink.Vars
-	}
-	if s.Uninstall != nil {
-		return s.Uninstall.Vars
-	}
-	return storage.OperationVariables{}
+	return (*storage.SiteOperation)(s).Vars()
```
Comment (a-palchikov) [inline lib/ops/ops.go:1109]: This was an alternative but not identical re-implementation.

File: tool/gravity/cli/clusterconfig.go [hunk 6]
```diff
@@ -189,10 +189,6 @@ func getConfigUpdater(localEnv, updateEnv *localenv.LocalEnvironment, operation
 	return updater, nil
 }
 
-func (r configInitializer) validatePreconditions(*localenv.LocalEnvironment, ops.Operator, ops.Site) error {
```
Comment (a-palchikov) [inline tool/gravity/cli/clusterconfig.go:192]: Removed as inherent noop - instead initializer can optionally implement a validator interface if required.

File: Makefile [hunk 7]
```diff
@@ -39,6 +39,8 @@ CURRENT_TAG := $(shell ./version.sh)
 GRAVITY_TAG := $(CURRENT_TAG)
 # Abbreviated gravity version to use as a build ID
 GRAVITY_VERSION := $(CURRENT_TAG)
+# Release of the gravity runtime application to builder installer with intermediate steps
+GRAVITY_INTERMEDIATE_RELEASE ?= 6.1.31
```
Comment (bernardjkim) [inline Makefile:43]: Will we want to keep this intermediate up to date with the latest 6.1 release? If so we should add an extra comment on when to update this value.

File: lib/builder/builder.go [hunk 8]
```diff
@@ -537,3 +590,65 @@ type GetRepositoryFunc func(*Builder) (string, error)
...
+		return nil, trace.BadParameter("expected a chart directory: %v", manifestPath)
+	}
+	chart, err := chartutil.Load(manifestPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return generateManifest(chart)
+}
+
+func filterUpgradePackageDependencies(packages []pack.PackageEnvelope) (result []pack.PackageEnvelope) {
```
Comment (bernardjkim) [inline lib/builder/builder.go:611]: nit: Can we add documentation on what packages are being/should be filtered?

File: Makefile (hunk 7 above)
Comment (a-palchikov) [reply Makefile:43]: This is a good question. That entry is mainly for the internal consumption when we build the installer tarball. The version choice is also sort of arbitrary - I went with the latest LTS release at the time I started working on it but there's no compelling reason to always use the latest.
I'll try to formulate a requirement for this as a comment.

File: lib/schema/manifest.go [hunk 9]
```diff
@@ -506,9 +513,19 @@ func (d Dependencies) ByName(names ...string) (*loc.Locator, error) {
 	return nil, trace.NotFound("dependencies %q are not defined in the manifest", names)
 }
 
+// Has returns true if dependencies have the specified package
```
Comment (bernardjkim) [inline lib/schema/manifest.go:516]: Godoc indicates bool return value, but function returns Locator.

File: lib/builder/syncer.go [hunk 10]
```diff
@@ -70,12 +69,8 @@ func newS3Syncer() (*s3Syncer, error) {
...
 // Sync makes sure that local cache has all required dependencies for the
 // selected runtime
-func (s *s3Syncer) Sync(builder *Builder, runtimeVersion *semver.Version) error {
-	tarball, err := s.hub.Get(loc.Locator{
-		Repository: defaults.SystemAccountOrg,
-		Name:       defaults.TelekubePackage,
-		Version:    runtimeVersion.String(),
-	})
+func (s *s3Syncer) Sync(ctx context.Context, builder *Builder, runtimeVersion semver.Version) error {
+	tarball, err := s.hub.Get(loc.Gravity.WithVersion(runtimeVersion))
```
Comment (r0mant) [inline lib/builder/syncer.go:73]: Have you made sure that open-source tele build still works with this change? I don't remember if "hub" used here supports "telekube <-> gravity" name translation.

File: lib/localenv/clusterenv.go [hunk 11]
```diff
@@ -71,6 +64,12 @@ func (r *LocalEnvironment) NewClusterEnvironment(opts ...ClusterEnvironmentOptio
 	if err != nil && !trace.IsNotFound(err) {
 		return nil, trace.Wrap(err)
 	}
+	ctx, cancel := context.WithTimeout(context.TODO(), defaults.AuditLogClientTimeout)
+	defer cancel()
+	auditLog, err := r.AuditLog(ctx)
+	if err != nil && !trace.IsNotFound(err) && !trace.IsAccessDenied(err) {
```
Comment (r0mant) [inline lib/localenv/clusterenv.go:70]: Why is access denied also being filtered out here?

File: lib/update/cluster/phases/bootstrap.go [hunk 12]
```diff
@@ -18,25 +18,233 @@ package phases
...
+		Manifest:       p.updateManifest,
+		RuntimePackage: server.Runtime.Update.Package,
+		Package:        &server.Runtime.Update.ConfigPackage,
+		Config:         p.existingClusterConfig,
+		Env:            p.existingEnviron,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = p.bootstrap.Packages.UpsertPackage(resp.Locator, resp.Reader,
```
Comment (r0mant) [inline lib/update/cluster/phases/bootstrap.go:195]: Is there any particular reason the planet config package is upserted, but the secrets package above is created (with already exists error ignored)?

File: tool/gravity/cli/update_test.go [hunk 13]
```diff
@@ -0,0 +1,87 @@
+package cli
```
Comment (r0mant) [inline tool/gravity/cli/update_test.go:1]: License header

File: tool/gravity/cli/update_test.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/loc"
+	"github.com/gravitational/gravity/lib/localenv"
+	"github.com/gravitational/gravity/lib/ops/opsservice"
+	"github.com/gravitational/gravity/lib/storage/keyval"
+
+	"gopkg.in/check.v1"
+)
+
+func TestCLI(t *testing.T) { check.TestingT(t) }
```
Comment (r0mant) [inline tool/gravity/cli/update_test.go:18]: Are tests in tool/ executed as a part of "make test"?

**Review by r0mant — APPROVED** [review]

File: tool/gravity/cli/update_test.go (hunk 14 above)
Comment (a-palchikov) [reply tool/gravity/cli/update_test.go:18]: Yes, but I just noticed that we don't run enterprise package tests - something to fix in future.

File: lib/update/cluster/phases/bootstrap.go (hunk 12 above)
Comment (a-palchikov) [reply lib/update/cluster/phases/bootstrap.go:195]: Hmm, I don't think there is any specific reason - I will update to use `UpsertPackage` API for the secrets too.

File: lib/localenv/clusterenv.go (hunk 11 above)
Comment (a-palchikov) [reply lib/localenv/clusterenv.go:70]: This does not work on regular nodes:
```
2020-08-04T17:30:00Z WARN [LOCAL]     Failed to create audit log. error:[
ERROR REPORT:
Original Error: *trace.AccessDeniedError access denied to perform action "create" on "user"
Stack Trace:
	/go/_/src/github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/services/role.go:1865 github.com/gravitational/gravity/vendor/github.com/gravitational/teleport/lib/services.RoleSet.CheckAccessToRule
	...
```

File: lib/schema/manifest.go (hunk 9 above)
Comment (a-palchikov) [reply lib/schema/manifest.go:516]: Good catch. Will remove since it is unused.

File: lib/builder/syncer.go (hunk 10 above)
Comment (a-palchikov) [reply lib/builder/syncer.go:73]: Good catch - I've changed the application to `telekube` for the oss version so this works.

**Pull request merged**